modules/tensor_mechanics/src/materials/MultiParameterPlasticityStressUpdate.C

```diff
@@ -0,0 +1,949 @@
+        getMaterialPropertyOld<std::vector<Real>>(_name_prepender + "plastic_internal_parameter")),
+    _yf(declareProperty<std::vector<Real>>(_name_prepender + "plastic_yield_function")),
+    _iter(declareProperty<Real>(_name_prepender +
+                                "plastic_NR_iterations")), // this is really an unsigned int, but
+                                                           // for visualisation i convert it to Real
+    _linesearch_needed(
+        declareProperty<Real>(_name_prepender + "plastic_linesearch_needed")) // this is really a
+                                                                              // boolean, but for
+                                                                              // visualisation i
+                                                                              // convert it to Real
+{
+  if (_definitely_ok_sp.size() != _num_sp)
+    mooseError("MultiParameterPlasticityStressUpdate: admissible_stress parameter must consist of ",
+               _num_sp,
+               " numbers");
+}
+
+void
+MultiParameterPlasticityStressUpdate::initQpStatefulProperties()
+{
+  _plastic_strain[_qp].zero();
+  _intnl[_qp].assign(_num_intnl, 0);
+  _yf[_qp].assign(_num_yf, 0);
+  _iter[_qp] = 0.0;
+  _linesearch_needed[_qp] = 0.0;
+}
+
+void
+MultiParameterPlasticityStressUpdate::propagateQpStatefulProperties()
+{
+  _plastic_strain[_qp] = _plastic_strain_old[_qp];
+  for (unsigned i = 0; i < _num_intnl; ++i)
+    _intnl[_qp][i] = _intnl_old[_qp][i];
+}
+
+void
+MultiParameterPlasticityStressUpdate::updateState(RankTwoTensor & strain_increment,
+                                                  RankTwoTensor & inelastic_strain_increment,
+                                                  const RankTwoTensor & rotation_increment,
+                                                  RankTwoTensor & stress_new,
+                                                  const RankTwoTensor & stress_old,
+                                                  const RankFourTensor & elasticity_tensor,
+                                                  const RankTwoTensor & /*elastic_strain_old*/,
+                                                  bool compute_full_tangent_operator,
+                                                  RankFourTensor & tangent_operator)
+{
+  initialiseReturnProcess();
+
+  if (_t_step >= 2)
+    _step_one = false;
+
+  // initially assume an elastic deformation
+  std::copy(_intnl_old[_qp].begin(), _intnl_old[_qp].end(), _intnl[_qp].begin());
+  _iter[_qp] = 0.0;
+  _linesearch_needed[_qp] = 0.0;
+
+  /**
+   * "Trial" value of stress_params that initialises the return-map process
+   * This is derived from stress = stress_old + Eijkl * strain_increment.
+   * However, since the return-map process can fail and be restarted by
+   * applying strain_increment in multiple substeps, _trial_sp can vary
+   * from substep to substep.
+   */
+  std::vector<Real> trial_sp(_num_sp);
```

Comment:
done.  i can see pros and cons with the readability of the code in various places with these changes, but you're a better coder than me, so i'll keep with your suggestion.  I have made the new variables private as i really don't want anyone else mucking around with the internals of the unfortunately complicated algorithms.